otaclient/app/update_stats.py

```diff
@@ -143,27 +132,38 @@ def collector(self):
                     for st in self._staging:
-                        if st.op in [
-                            RegProcessOperation.OP_COPY,
-                            RegProcessOperation.OP_DOWNLOAD,
-                            RegProcessOperation.OP_LINK,
-                        ]:
-                            _suffix = st.op
-                            staging_storage.regular_files_processed += 1
-                            staging_storage[f"files_processed_{_suffix}"] += 1
-                            staging_storage[f"file_size_processed_{_suffix}"] += st.size
-                            staging_storage.add_elapsed_time(
-                                f"elapsed_time_{_suffix}", st.elapsed_ns
-                            )  # in nano-seconds
-                            if st.op == RegProcessOperation.OP_DOWNLOAD:
-                                staging_storage[
-                                    f"errors_{RegProcessOperation.OP_DOWNLOAD.value}"
-                                ] += st.errors
-                                staging_storage.download_bytes += st.download_bytes
-                        elif st.op == RegProcessOperation.OP_DOWNLOAD_FAILED_REPORT:
-                            staging_storage[
-                                f"errors_{RegProcessOperation.OP_DOWNLOAD.value}"
-                            ] += st.errors
+                        _op = st.op
+                        if _op == RegProcessOperation.DOWNLOAD_REMOTE_COPY:
+                            # update download specific fields
+                            staging_storage.downloaded_bytes += st.downloaded_bytes
+                            staging_storage.downloaded_files_num += 1
+                            staging_storage.downloaded_files_size += st.size
+                            staging_storage.downloading_errors += st.download_errors
+                            staging_storage.downloading_elapsed_time.add_nanoseconds(
+                                st.elapsed_ns
+                            )
+                            # as remote_delta, update processed_files_*
+                            staging_storage.processed_files_num += 1
+                            staging_storage.processed_files_size += st.size
+                        elif _op == RegProcessOperation.DOWNLOAD_ERROR_REPORT:
+                            staging_storage.downloading_errors += 1
+                        elif _op == RegProcessOperation.PREPARE_LOCAL_COPY:
+                            # update delta generating specific fields
+                            staging_storage.delta_generating_elapsed_time.add_nanoseconds(
+                                st.elapsed_ns
+                            )
+                            # as keep_delta, update processed_files_*
+                            staging_storage.processed_files_size += st.size
+                            staging_storage.processed_files_num += 1
```

Comment:
Yes, `downloading`, `preparing_local_copy` and `applying_delta` both contributes to `processe_file_*`, so there is a special treatment in `applying_delta` to prevent double counting `processed_files_size/processed_files_num`.

The logic is that, for regular files with same hash, there is at most one copy either being prepared by downloading, or by local_copy, and the left over copies are prepared by copying/hardlinking the first copy.
https://github.com/tier4/ota-client/blob/285a580fc28453060d8c3444ab9ec4b63a9c8b13/otaclient/app/create_standby/rebuild_mode.py#L187-L191

Reply:
I could understand the logic but it is better to add comment to prevent double counting for this block.
rebuild_mode.py knows update_stats.py internal implementation, but it is better for update_stats.py to have this note as a comment.

Reply:
Updated at [3a2a07f](https://github.com/tier4/ota-client/pull/205/commits/3a2a07fd9ff63dfe42043126c752ea570c61d757).

Reply:
Updated in [9fcbc11](https://github.com/tier4/ota-client/pull/205/commits/9fcbc11dc1f4320385282b6ec1c925a0f8735fd6).